src/NuGet.Clients/PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs

```diff
@@ -0,0 +1,135 @@
+
+            MigrateRuntimes(packageSpec, buildProject);
+
+            RemoveProjectJsonReference(buildProject, projectJsonFilePath);
+
+            CreateBackup(project,
+                projectJsonFilePath);
+        }
+
+        private static async Task MigrateDependencies(LegacyCSProjPackageReferenceProject project, PackageSpec packageSpec)
+        {
+            if (packageSpec.TargetFrameworks.Count > 1)
+            {
+                throw new InvalidOperationException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.Error_MultipleFrameworks,
+                        project.MSBuildProjectPath));
+            }
+
+            var dependencies = new List<LibraryDependency>();
+            foreach (var targetFramework in packageSpec.TargetFrameworks)
+            {
+                dependencies.AddRange(targetFramework.Dependencies);
+            }
+
+            dependencies.AddRange(packageSpec.Dependencies);
+            foreach (var dependency in dependencies)
+            {
+                var includeFlags = dependency.IncludeType;
+                var privateAssetsFlag = dependency.SuppressParent;
+                var metadataElements = new List<string>();
+                var metadataValues = new List<string>();
+                if (includeFlags != LibraryIncludeFlags.All)
+                {
+                    metadataElements.Add("IncludeAssets");
+                    metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(includeFlags).Replace(',',';'));
+                }
+
+                if (privateAssetsFlag != LibraryIncludeFlagUtils.DefaultSuppressParent)
+                {
+                    metadataElements.Add("PrivateAssets");
+                    metadataValues.Add(LibraryIncludeFlagUtils.GetFlagString(privateAssetsFlag).Replace(',',';'));
+                }
+
+                await project.InstallPackageWithMetadataAsync(dependency.Name, dependency.LibraryRange.VersionRange, metadataElements, metadataValues);
+            }
+        }
+
+        private static void MigrateRuntimes(PackageSpec packageSpec, Microsoft.Build.Evaluation.Project buildProject)
+        {
+            var runtimes = packageSpec.RuntimeGraph.Runtimes;
+            var supports = packageSpec.RuntimeGraph.Supports;
+            var runtimeIdentifiers = new List<string>();
+            var runtimeSupports = new List<string>();
+            if (runtimes != null && runtimes.Count > 0)
+            {
+                runtimeIdentifiers.AddRange(runtimes.Keys);
+                
+            }
+
+            if(supports != null && supports.Count > 0)
+            {
+                runtimeSupports.AddRange(supports.Keys);
+            }
+
+            var union = string.Join(";", runtimeIdentifiers.Union(runtimeSupports));
+            NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
```

Comment:
I gather that this method is ultimately called from your public entrypoint that is already using JTF.Run. Please try to avoid nesting them. Instead, let this method be async so that you can naturally `await` within it. In fact I think this method's caller is already async so it should be cheap to make this method async too.
There are corner cases where a JTF.Run nested within another can *sometimes* cause deadlocks due to TPL Task continuation inlining. While it's rare, it's very hard to predict so it's best to keep to the safe pattern of using JTF.Run just from your public entrypoints that must be synchronous (and let them be async where possible).